File: ios/ios/Common/WrapperScreen.swift
import shared
import Combine
import SwiftUI

struct WrapperScreen<VM, ViewState, Action, Event, UI> : View where VM: ViewModel<ViewState, Action, Event>, UI: View {
    
    private let viewModel: VM
    private let content: (ViewState, @escaping (Event) -> Void) -> UI
    private let onAction: (Action?) -> Void

    init(
        viewModel: VM,
        content: @escaping (ViewState, @escaping (Event) -> Void) -> UI,
        onAction: @escaping (Action?) -> Void)
    {
        self.viewModel = viewModel
        self.content = content
        self.onAction = onAction
    }
    
    
    var body: some View {
        ObservingView(statePublisher: toPublisher(viewModel.viewStates())) { state in
            content(state) { event in
                viewModel.obtainEvent(viewEvent: event)
            }
        }.onReceive(toPublisher(viewModel.viewActions())) { action in
            onAction(action)
        }
    }
}

File: android/src/main/java/ua/ukma/edu/programistich/kmmtwitter/android/common/WrapperScreen.kt
package ua.ukma.edu.programistich.kmmtwitter.android.common

import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisallowComposableCalls
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.remember
import ua.ukma.edu.programistich.kmmtwitter.common.viewmodel.ViewModel

@Composable
fun <T : ViewModel<State, Action, Event>, State, Action, Event> WrapperScreen(
    viewModelFactory: @DisallowComposableCalls () -> T,
    onAction: (Action?) -> Unit,
    content: @Composable (State, (Event) -> Unit) -> Unit
) {
    val viewModel = remember { viewModelFactory() }
    val viewState = viewModel.viewStates().collectAsState()
    val viewAction = viewModel.viewActions().collectAsState(initial = null)
    val obtainEvent = { event: Event ->
        viewModel.obtainEvent(event)
    }

    onAction(viewAction.value)

    content(viewState.value) {
        obtainEvent(it)
    }

    DisposableEffect(Unit) {
        onDispose(viewModel::clear)
    }
}
